[장바구니 협업 Step 2] 참새(김규) 미션 제출합니다.

WaiNaat opened the pull request:
안녕하세요 포코! 참새입니다.
원래는 주말 전에 PR 드리고 싶었는데 미션이 생각보다 잘 안풀렸어요😥 부족한 부분이 너무 많습니다.
전반적으로 리팩터링이 덜 되어 코드가 너무 지저분하네요ㅠㅠ

## 배포 링크 

[장바구니 협업 미션](https://wainaat.github.io/react-shopping-cart-prod)
- 알려진 버그!!!! 첫 접속 이후 처음 서버를 바꾸면 장바구니가 계속 로딩 상태에 있어요😭 다른 서버로 바꾸면 정상 작동합니다.
- 아직 백엔드 서버가 완벽하게 구현된 게 아니라 작동이 안 될 수도 있어요

[스토리북](https://wainaat.github.io/react-shopping-cart-prod/storybook)

## 고민했던 점들

### 훅 분리

기존의 `useCartService`는 너무 많은 일을 하고 있었습니다. 장바구니 수량 변경 api 호출, 전역 장바구니 상태, 서버로부터 전체 장바구니 받아오는 api 호출을 전부 수행하고 있었는데요. 이 부분을 쪼갰습니다.

`useCartService`의 이름을 `useCartItemApi`로 바꾸고, 장바구니에 한 개의 상품을 넣고 빼는 로직을 담당하도록 하였습니다.
전역 장바구니 상태 관리는 `getCartStateController`, 장바구니 받아오는 api는 `fetchCartItems`라는 recoil selector가 할 수 있도록 분리했어요.

마음 같아서는 fetch 관련 로직을 더 잘 분리할 수 있을 것 같은데 시간과 실력 문제로 이번 PR에서는 제대로 하지 못한 상태입니다.. 리액트 공식 문서에 `Describing the UI`를 읽으니까 결국 리액트 컴포넌트는 UI를 보여주는 게 가장 큰 역할이라는 생각이 듭니다. 그래서 로직들은 훅으로 분리했으면 좋겠는데 제대로 하지 못해서 너무 아쉬워요

### 통신 에러 처리

Suspense와 ErrorBoundary를 이용해서 데이터 불러오기에 대한 실패 처리를 최대한 해보고자 하였습니다.
그래서 장바구니나 주문 목록, 상품 목록 불러오기는 잘 막히는 것 같아요!

하지만 수량을 정하는 Counter에서의 통신 실패 처리는 아직 하지 못한 상태입니다.
낙관적 업데이트를 하고 있어서 실패했을 경우 뒤로 돌리는 로직을 넣고 싶은데 시간 문제로 건드리지도 못했네요 😥

Suspense를 잘 사용하고 있는지도 의문입니다. Recoil의 비동기 atom이나 selector를 이용하면 자동적으로 suspense가 걸 수 있다는 건 알겠지만 promise를 throw하는 방식이 맞는 건지는 잘 모르겠어요. 지금은 useFetch()에서 getData()를 호출할 때 promise를 throw하는 방식을 이용하고 있습니다. 이 부분도 더 깔끔하게 하고 싶은데 아직 경험의 부족으로 생각이 안 나네요..

### msw

처음에는 '백엔드 서버가 오는데 msw가 필요할까?'라는 생각을 하고 step2에 대한 msw를 만들지 않았습니다. 하지만 이렇게 하니 저는 백엔드를 기다려야만 하더라구요. 급하게 깨달음을 얻고 step2에 대한 msw를 만들었습니다. 확실히 msw를 만들면서 하니까 확인하기가 너무 좋은 것 같아요!

지금은 db 대신에 로컬스토리지를 사용하도록 구현했습니다. 로컬스토리지에 `MSW_`라는 이름으로 저장된 값들은 다 MSW에서만 접근하는 값들이에요.

저는 사실 MSW가 너무 마음에 듭니다😅 기회가 된다면 앞으로도 계속 msw를 만들어가면서 개발을 하는 게 좋을 것 같아요.

### 주문 목록 -> 주문 상세 페이지 이동에서의 값 전달 방법

저희는 api에서 주는 주문 목록이 주문 상세정보의 배열입니다. 그래서 저는 주문 목록 -> 주문 상세로 이동할 때 굳이 새로운 요청을 보낼 필요가 없다고 판단했어요. 그래서 어떤 방식으로 주문 상세 정보를 전달할지 고민이 되었습니다.

구현은 주문 목록에서 주문 상세로 이동할 때 recoil atom에 상세 정보를 담고, 주문 상세 페이지에서 그 atom을 읽어서 출력하는 방식을 사용했습니다. 하지만 이 방법이 맞나? 라는 의문은 들어요.

제가 생각한 방법은 recoil atom 말고도 하나가 더 있었습니다. navigate할 때 history api의 state로 주문 상세를 넘기는 방법이었는데요. 이렇게 하면 전역 상태를 이용하지 않는다는 장점이 있었습니다. 하지만 주문 상세에는 사실 길이 제한이 없어서, 직렬화-역직렬화 비용을 생각해서 이 방법을 사용하지는 않았어요.

---

어쩌다보니 PR이 아니라 반성문처럼 되어버린 것 같아요ㅋㅋㅋ
레벨2 마지막 PR도 잘 부탁드립니다!!

WaiNaat committed:
refactor: 백엔드 서버 추가

WaiNaat committed:
chore: eslint 설정

WaiNaat committed:
style: eslint 설정에 맞게 수정

WaiNaat committed:
fix: 장바구니 상품 표시 안되는 오류 수정

WaiNaat committed:
refactor: 장바구니 첫 추가시 await

장바구니에 추가 후 cartItemId를 받아와야 해서 대기

WaiNaat committed:
refactor: 불필요한 state 삭제

WaiNaat committed:
feat: 장바구니 물품 1개의 수량 조절하는 selector

WaiNaat committed:
refactor: useCartService 훅 관심사 분리

- 관련 커밋 0292312
- 기존: 장바구니 수량 변경 api 호출과 전역상태 관리를 동시에 실행
- 현재: 관련 커밋에서 전역상태 관리를 분리하여 api통신만 남김

WaiNaat committed:
feat: productId로 cartItemId 가져오는 selector

WaiNaat committed:
refactor: useCartService 관심사 분리

- 관련 커밋: 17433ea
- useCartService에서는 api 통신만 하기를 원함
- 따라서 cartItemId를 가져오는 로직을 제거

WaiNaat committed:
refactor: 불필요한 state 삭제

WaiNaat committed:
feat: 장바구니 길이 가져오는 selector

WaiNaat committed:
feat: 장바구니를 불러오는 selector

WaiNaat committed:
refactor: useCartService 관심사 분리

- 관련 커밋 287060d
- 장바구니 fetch 로직을 selector로 분리
- useCartService는 장바구니 수량 조절 api만 담당하게끔 변경
- useEffect로 서버명 변경 시 자동 fetch
- 장바구니 수량 조절을 recoil로 관리하여 cartLoadingState 불필요

WaiNaat committed:
refactor: 서버 변경 시 체크박스 초기화

WaiNaat committed:
refactor: 불필요한 콘솔 출력 삭제

WaiNaat committed:
refactor: 서버 선택 ui 컴포넌트 분리

WaiNaat committed:
refactor: 장바구니 이동 버튼 컴포넌트 분리

WaiNaat committed:
design: 서버 선택 컴포넌트 UI 변경

WaiNaat committed:
refactor: 불필요한 suspense 삭제

WaiNaat committed:
refactor: 장바구니 페이지 suspense 적용

WaiNaat committed:
refactor: Counter에 suspense 적용

- 서버 변경 시 두 번 깜빡이는 문제 개선

WaiNaat committed:
chore: msw deps로 이동

WaiNaat committed:
refactor: 스토리북 전용 msw 적용

WaiNaat committed:
refactor: 불필요한 파일 삭제

WaiNaat committed:
chore: react-icons 패키지 설치

WaiNaat committed:
feat: 쿠폰 컴포넌트 구현

WaiNaat committed:
refactor: svg 아이콘 react-icons로 수정

WaiNaat committed:
test: 스토리북 코드 수정

WaiNaat committed:
fix: typo

WaiNaat committed:
feat: 쿠폰 받기 페이지 구현

WaiNaat committed:
refactor: useFetch error 로직 변경

WaiNaat committed:
refactor: Errorboundary 임시 재시도 버튼 추가

WaiNaat committed:
refactor: 쿠폰 목록 suspense 적용

WaiNaat committed:
refactor: 서버 요청 경로 상수화

WaiNaat committed:
feat: 쿠폰 받기 배너 추가

WaiNaat committed:
refactor: 서버 선택 ui 디렉토리 변경

- 헤더에 있지만 사실 꼭 헤더에 있을 필요는 없음
- 헤더의 다른 기능들과는 달리 이질적임

WaiNaat committed:
refactor: 서버 선택에 msw 추가

WaiNaat committed:
feat: msw 쿠폰 발급 기능 추가

WaiNaat committed:
refactor: 쿠폰 props 변경

- 할인 정보를 받아서 자동으로 표시하도록 수정

WaiNaat committed:
refactor: 쿠폰 API 명세 수정으로 인한 변경

WaiNaat committed:
feat: 장바구니에서 쿠폰 선택 기능 추가

WaiNaat committed:
refactor: useFetch auth 기능 변경

WaiNaat committed:
refactor: 서버 변경 시 선택된 쿠폰 초기화

WaiNaat committed:
fix: 쿠폰 적용시 overflow 현상 해결

WaiNaat committed:
design: 쿠폰 선택 UI

WaiNaat committed:
test: 쿠폰 선택 컴포넌트 테스트

WaiNaat committed:
feat: msw 주문하기 api 기능 구현

WaiNaat committed:
feat: 주문하기 버튼 기능 구현

WaiNaat committed:
feat: 주문 중일 때 스피너 추가

WaiNaat committed:
feat: 주문 성공 시 완료 페이지로 이동

WaiNaat committed:
refactor: 쿠폰 발급 성공 시 안내 문구 추가

WaiNaat committed:
refactor: 서버 변경 시 메인페이지로 리다이렉트

WaiNaat committed:
design: 쿠폰 선택 ui 개선

WaiNaat committed:
refactor: msw 새로운 상품 추가

WaiNaat committed:
feat: 주문 목록의 상품 1개 컴포넌트

WaiNaat committed:
feat: 주문 상세보기 컴포넌트

WaiNaat committed:
feat: 주문 목록 페이지 구현

WaiNaat committed:
feat: 주문 상세 페이지 구현

WaiNaat committed:
refactor: 주문 성공 페이지에서 주문 목록 이동 가능

WaiNaat committed:
design: 주문 상세 최대 가로길이 설정

WaiNaat committed:
design: 쿠폰 선택창 크기 변경

WaiNaat committed:
refactor: api 명세에 따른 임시 코드 구현

WaiNaat committed:
refactor: 훅 이름 직관적으로 변경

WaiNaat committed:
fix: 서버에서 배송비 안보내주는 경우 임시 수정

WaiNaat committed:
chore: 자동 배포 설정

WaiNaat committed:
chore: 자동 배포를 위한 eslint 설정 변경

WaiNaat committed:
chore: gh-pages 배포 msw 설정

WaiNaat committed:
chore: 자동 배포 취소

WaiNaat committed:
feat: 페이지 전용 ErrorBoundary 구현

WaiNaat committed:
fix: 로컬호스트에서 msw 돌아가게 설정

WaiNaat committed:
feat: 장바구니 쿠폰 ErrorBoundary

WaiNaat committed:
fix: typo

WaiNaat committed:
feat: 헤더 장바구니 개수 ErrorBoundary

WaiNaat committed:
refactor: 쿠폰 발급 실패 안내 추가

WaiNaat committed:
refactor: 불필요한 주석 삭제

WaiNaat committed:
fix: 스토리북 msw 오류 수정

pocojang commented on .eslintrc.json:
`airbnb`를 제외하고는 `CRA` 기본과 중복이 꽤 많지 않을까요?

WaiNaat replied:
https://github.com/facebook/create-react-app/tree/main/packages/eslint-config-react-app
여기 설정과 비교해서 중복되는 부분은 없애겠습니다.

pocojang commented on package.json:
```diff
@@ -3,8 +3,10 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "msw": "^1.2.1",
```
CRA & 정적 빌드라 큰 영향은 없겠지만 위치가 좀 애매하군요..?

WaiNaat replied:
원래는 devDeps에 있었고 일반적으로 서비스하는 어플리케이션이라면 devDeps에 있어야 한다고 생각합니다. 하지만 지금은 배포하는 곳에서 msw를 사용해서 deps로 옮겼어요

pocojang commented on src/components/ServerSelector/ServerSelector.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  useEffect(() => {
+    resetCart();
+    navigate('/', { replace: true });
```
`replace: true` 무엇때문에 설정하셨을까요?

WaiNaat replied:
서버를 바꾸는 행동이 로그아웃/로그인보다는 더 중요한? 일이라는 생각이 들었습니다. 지금은 서버 교체를 마치 유저 교체처럼 사용하고 있지만, 실제로는 개발서버/배포서버 전환처럼 아예 다른 목적을 가진 서버 사이의 이동에 사용할 것 같았어요. 그래서 전환 이후에는 뒤로가기로 이전에 접근하기 힘들도록 replace를 설정했습니다. 근데 두 단계 이전의 기록들은 남아있어서 무의미한 것 같기도 하네요..

pocojang replied:
굿!! 이유가 있으셨군요

pocojang commented on src/components/ServerSelector/ServerSelector.tsx:
```diff
@@ -0,0 +1,57 @@
+    navigate('/', { replace: true });
+  }, [serverName]);
+
+  const handleServerNameSelectChange: React.ChangeEventHandler<HTMLSelectElement> = (event) => {
```
흠.. 이게 여기 있는게 맞을지 계속 고민되네요

WaiNaat replied:
급한 마음에 msw를 끄고 켜는 로직을 여기에 넣게 되었습니다..
생각해보니 `참새MSW`에서 다른 서버로 이동할 때 msw를 끌 필요가 있는지 의문이 드네요. 현재 msw의 역할은 개발 도우미뿐 아니라 실제 배포에서 하나의 백엔드 서버처럼 동작합니다. 백엔드 서버 이름이 바뀐다고 이전 서버가 꺼지지 않는다는 점을 생각하면 굳이 서버가 바뀔 때 msw를 켜고 끌 필요는 없는 것 같아요.
그래서 리팩터링은 msw 관련 로직을 여기서 지우는 방향으로 진행하겠습니다!

pocojang commented on src/components/cart/CartContents/CartContents.tsx:
```diff
@@ -0,0 +1,98 @@
+const CouponFallbackDiv = styled.div`
+  padding: 20px 30px;
+
+  border: 1px solid #dddddd;
+  color: #333333;
```
컬러 관리를 더 해보시면 좋겠습니다 :)

WaiNaat replied:
색상값들은 상수화하여 사용하도록 하겠습니다!

pocojang commented on src/components/cart/CartContents/CartContents.tsx:
```diff
@@ -0,0 +1,98 @@
+  return cartLength ? (
+    <Contents>
+      <CartList />
+      <PaymentDiv>
```
`PaymentDiv` 네이밍이 .. 흠.. 애매한데요? 어떠세요?

WaiNaat replied:
혹시 `Div` 부분 때문인가요? 제가 스타일드 컴포넌트를 사용하면서 느꼈던 불편한 점이 두 가지였는데요. 딱 봤을 때 이게 스타일드 컴포넌트인지 생각해야 한다는 점과 html의 어떤 태그인지 알기 힘들다는 점이었습니다. 그래서 그 점을 명시해 주고자 이름에 div를 붙이는 방법을 적용해 보았어요. 혹시 많이 사용하지 않는 방법인가요??

(다른 스타일드 컴포넌트들은 이번 협업 미션이 아니라 그 이전 미션에서 작업했었던 내용이라 전부 고치기에는 시간이 모자랄 것 같아서 건드리지 못했습니다)

pocojang replied:
대중적이다 기준으로 보기에는 애매하고 올바른 방향은 아닌 것 같네요 ㅎㅎ

pocojang commented on src/components/cart/CartContents/CartContents.tsx:
```diff
@@ -0,0 +1,98 @@
+  const [selectedCoupon, setSelectedCoupon] = useState<CouponInfo | null>(null);
+  const cartLength = useRecoilValue(getCartLength);
+
+  const handleLinkButtonClick = () => {
```
https://blog.shiren.dev/2020-07-27-1

WaiNaat replied:
이벤트 핸들러는 이벤트 객체를 해석해서 원하는 정보를 뽑아낸 후 특정 메서드들을 부르는 역할이군요! 이벤트 객체에 의존하지 않는 방향으로 함수를 작성하고 네이밍 역시 해당 함수가 하는 일을 표현하는 방향으로 개선하겠습니다.

pocojang commented on src/components/cart/CartContents/CartContents.tsx:
```diff
@@ -0,0 +1,98 @@
+import CartList from '../CartList/CartList';
+import PaymentAmount from '../PaymentAmount/PaymentAmount';
+import getCartLength from '../../../globalState/selectors/getCartLength';
+import CouponInfo from '../../../types/coupon';
```
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

WaiNaat replied:
타입임을 명시해주면 알아보기도 좋고 런타임 코드에서도 무조건 빼 주는군요 감사합니다!!

pocojang commented on src/components/cart/CartItem/CartItem.tsx:
```diff
@@ -1,52 +1,57 @@
 
 const CartItem = ({ cartItem }: CartItemProps) => {
-  const { id, quantity } = cartItem;
+  const { id: cartItemId, quantity } = cartItem;
```
오 센스 좋아요 👍

pocojang commented on src/components/cart/CartItem/CartItem.tsx:
```diff
@@ -1,52 +1,57 @@
 
-    deleteCartItem(id);
-    deleteCheckedItem(id);
+    deleteCartItem(cartItemId);
+    deleteCheckedItem(cartItemId);
+    cartContoller.delete();
```
동기적으로 순서가 지켜져야하는 로직일까요?

WaiNaat replied:
지금 코드에서는 순서가 상관없습니다.
하지만 deleteCartItem이 api 호출하는 부분이라 실패했을 때를 고려하는 코드를 추가한다면 순서가 유의미할 것 같아요

pocojang commented on src/components/cart/PaymentAmount/PaymentAmount.tsx:
```diff
@@ -1,35 +1,121 @@
+import LoadingSpinner from '../../common/LoadingSpinner/LoadingSpinner';
+
+interface PaymentAmountProps {
+  coupon?: CouponInfo | null;
```
타입이 뭔가.. 이상하지 않나요? 🤔

WaiNaat replied:
그러네요😥
optional로 하거나 null을 허용하거나 둘 중 하나만 했어야 했는데 둘 다 써버린 것 같습니다ㅠㅠ

pocojang commented on src/components/cart/PaymentAmount/PaymentAmount.tsx:
```diff
@@ -1,35 +1,121 @@
+
+  const navigate = useNavigate();
+
+  const discountedPrice = (() => {
+    if (!coupon) return 0;
+    if (coupon.type === 'percent') return Math.ceil(paymentAmount * (coupon.amount / 100));
+    return coupon.amount;
+  })();
```
`IIFE`를 쓰신 이유가 궁금합니다!

WaiNaat replied:
현재 쿠폰 정책상 분기점이 `쿠폰이 없는 경우`, `퍼센트 할인 쿠폰인 경우`, `고정 금액 할인 쿠폰인 경우` 세 가지로 나뉩니다.

```javascript
let discountedPrice = 0;
if (coupon) {
  discountedPrice = coupon.type === 'percent' ? Math.ceil(paymentAmount * (coupon.amount / 100)) : coupon.amount;
}
```
함수를 쓰지 않는다면 이런 느낌으로 let을 쓰거나 삼항 연산자 두 개를 쓰거나 하는 다른 방법들이 있었는데 저는 지저분하다는 생각이 들었어요. 그래서 early return을 이용하기 위해 즉시 실행 함수를 사용했습니다.

pocojang replied:
흠.. 그렇군요
사실 `IIFE`가 어떤 영향을 주는지 더 테스트해보시면 좋겠는걸요.. 큰 영향이 없는 행위로 보입니다

pocojang commented on src/components/common/Header/Header.tsx:
```diff
@@ -1,51 +1,37 @@
-  const handleLogoClick = () => {
-    navigate('/');
-  };
+  const gotoMain = () => navigate('/');
```
goToMain..?

WaiNaat replied:
어디로 가는 건지를 알려주고 싶었는데 너무한 네이밍이었을까요..?
일단은 별도의 전처리 없이 단순히 페이지 이동만 하는 로직을 위한 핸들러는 불필요하다고 느껴져서 전부 `<Link>` 컴포넌트로 바꿀 생각입니다

pocojang replied:
아녀 ㅎㅎ 네이밍 오타 아니냐는 질문이였습니다 ㅎㅎ

pocojang commented on src/components/common/Header/Header.tsx:
```diff
@@ -1,51 +1,37 @@
-    HTMLSelectElement
-  > = (event) => {
-    const selectedServerName = event.target.value;
+  const gotoCart = () => navigate('/cart');
```
goToCart...?

pocojang commented on src/components/coupon/CounponList/CouponList.tsx:
```diff
@@ -0,0 +1,67 @@
+  const { getData } = useFetch<{ coupons: CouponInfo[] }>(couponsUrl);
+  const data = getData();
+
+  const makeDownloadCouponFunction = (couponId: number) => async () => {
```
오 `Closure`를 즐겨 사용하시네요
어떤 이점이 있을까요?

WaiNaat replied:
저는 크게 두 가지 이점이 있다고 생각합니다!

1. 함수 공장 역할
이 코드처럼 내부 로직이 같은 함수를 공장처럼 만들어낼 수 있습니다. 이렇게 하면 만들어진 async 함수를 사용하는 측에서는 `couponId`에 대한 정보를 아예 몰라도 된다는 장점이 있다고 생각해요.
2. 함수만을 위한 비밀 저장소 부여
지금은 바로 함수를 리턴하지만 그 사이에 변수같은 걸 넣어서 함수의 호출 횟수를 세거나, debounce를 적용하는 등 해당 함수만 접근 가능한 변수들을 선언할 수 있습니다

pocojang replied:
흠 그렇군요!
기준점은 좋습니다
다만 효과가 있는지 꼭 검증하시길 바랍니다

pocojang commented on src/components/coupon/CounponList/CouponList.tsx:
```diff
@@ -0,0 +1,67 @@
+      body: JSON.stringify({ couponId }),
+    });
+
+    if (Math.floor(response.status / 100) === 5) {
+      alert('서버 문제로 쿠폰 발급에 실패하였습니다.');
+      return;
+    }
+
+    if (Math.floor(response.status / 100) === 4) {
+      alert('이미 발급받은 쿠폰입니다!');
+      return;
+    }
```
`response.status`가.. 제가 아는 그 `status`가 맞을가요..?

WaiNaat replied:
네... 예외 처리를 하지 못해서 급하게 두 가지 경우만 나눴습니다 😥

pocojang commented on src/components/order/OrderDetail/OrderDetail.tsx:
```diff
@@ -0,0 +1,176 @@
+            <Link onClick={setOrderDetails} to="/order-details">
+              상세보기 <GrFormNextLink size="23px" color="#333333" />
+            </Link>
+          ) : null}
```
`<></>` vs `null` vs `false` 차이를 아시나요?

WaiNaat replied:
`fragment`는 dom에 렌더링되지는 않지만 element들을 감싸는 데 쓰일 수 있어서 렌더링 시도는 한다고 알고 있습니다.
`null`과 `false`는 둘 다 렌더링 시도를 하지 않습니다. 하지만 `false`는 컴포넌트 렌더 함수에서 `return false;`로 빠져나올 때는 그 컴포넌트를 unmount 시키고, `null`은 그렇지 않다는 차이가 있습니다.

지금처럼 특정 부분에 대해서만 삼항 연산자로 null이나 false를 줬을 때의 차이는 잘 모르겠네요..

pocojang commented on src/components/order/OrderedProduct/OrderedProduct.tsx:
```diff
@@ -0,0 +1,73 @@
+
+  return (
+    <OrderedProductDiv>
+      <Image src={imageUrl} size="medium" alt="" />
```
`alt=""` 이건 뭘까요..?

WaiNaat replied:
처음에는 alt를 그냥 상품명으로 했었는데요. 이 경우에는 이미지 바로 옆에 해당 상품명이 있어서 adjacent text alternative가 있는 [decorative image](https://www.w3.org/WAI/tutorials/images/decorative/)와 다를 바가 없다고 느꼈습니다. 스크린 리더로 읽을 때도 '샐러드 이미지,  샐러드' 이런 식으로 alt를 읽고 이름을 읽는 같은 내용의 반복이 있었어요

그렇다고 이 이미지를 decorative하다고 판단하자니 저희가 물건을 살 때 이름만 듣고 사는 경우가 없다는 생각이 들었어요. 그러면 그 [상품을 '묘사'한 내용을 alt에 적어야 하는데](https://www.w3.org/WAI/tutorials/images/informative/#example-2-images-used-to-supplement-other-information) 이 부분을 프론트엔드가 일일이 적을 수는 없겠더라구요. 처음에 db에 등록할 때부터 alt를 같이 넣거나, 이미지 캡셔닝 같은 외부 기술을 끌어와야 살릴 수 있겠다는 생각을 했습니다.

그래서 우선은 alt를 상품명으로 했을 때 생기는 무의미한 반복이라도 없애보자는 생각에 `alt=""`를 사용했습니다

pocojang replied:
흠... 그렇군요 아무리 그래도 빈 값을 할당하는 것은 애매한데요..?

pocojang commented on src/errorHandler/ProductsErrorBoundary.tsx:
```diff
@@ -0,0 +1,46 @@
+    return hasError ? (
+      <ErrorDiv>
+        {message}
+        <RetryButton onClick={() => this.retry()}>다시 시도하기</RetryButton>
```
```suggestion
        <RetryButton onClick={this.retry}>다시 시도하기</RetryButton>
```

pocojang commented on src/hooks/api/useFetch.ts:
재활용이 잘 되지 않아 아쉽군요..

WaiNaat replied:
😥최대한 고쳐 보겠습니다..

pocojang commented on src/index.tsx:
```diff
@@ -6,6 +6,24 @@ import App from './App';
+import OrderListPage from './pages/OrderListPage';
+import OrderDetailPage from './pages/OrderDetailPage';
+
+(async () => {
+  if (window.location.pathname === '/react-shopping-cart-prod') {
+    window.location.pathname += '/';
+    return;
+  }
+
+  await worker.start({
+    serviceWorker: {
+      url: `${process.env.PUBLIC_URL}/mockServiceWorker.js`,
+    },
+  });
+})();
```
이 코드들은 개발환경 분기가 없어도 될까요?

WaiNaat replied:
실제 서비스라면 분기 처리를 해야겠지만 지금은 msw를 또다른 백엔드 서버처럼 사용하고 있어요. 깃허브 페이지로 배포하고 나서 백엔드 서버에 문제가 생겼을 경우 하나라도 돌아가게 하고 싶어서 따로 개발 환경 처리를 하지 않았습니다.

pocojang replied:
아하 의도가 있었군요
알겠습니다!

pocojang commented on src/types/coupon.ts:
```diff
@@ -0,0 +1,8 @@
+  name: string;
+}
+
+export default CouponInfo;
```
`default`가 맞을까요?

WaiNaat replied:
저는 개인적으로 default export를 선호해서 파일 안에 하나만 있으면 습관적으로 default export를 합니다.
근데 포코의 질문을 듣고 생각해보니 타입의 경우에는 default로 하면 쓰는 쪽에서 이름을 자유롭게 정할 수 있다는 점이 문제가 되겠더라구요. 그 타입을 쓰는 여러 모듈을 한 군데에 모아 두면 혼란스럽겠다는 생각이 들었어요

`CouponInfo`와 아래 `OrderInfo` 모두 named export로 바꾸겠습니다!

pocojang replied:
그래도 모두 자신만의 기준이 있으시군요 ㅎㅎ

pocojang commented on src/pages/OrderCompletePage.tsx:
```diff
@@ -0,0 +1,56 @@
+  const { orderId } = location.state ?? {};
+
+  useEffect(() => {
+    if (Number.isNaN(Number(orderId))) {
```
조건 로직이 특이한데.. 이게 맞을까요?

WaiNaat replied:
location.state에 저장한 orderId값이 문자열인 줄 알아서 이런 처리를 했었는데 숫자였네요...
```suggestion
    if (orderId === undefined) {
```
이렇게 바꾸면 될 것 같습니다

pocojang commented on src/pages/OrderCompletePage.tsx:
```diff
@@ -0,0 +1,56 @@
+const OrderCompletePage = () => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const { orderId } = location.state ?? {};
```
빈 객체가 할당되면.. `orderId`는 어떻게 될까요..?

WaiNaat replied:
`undefined`가 되는 것으로 알고 있습니다!

pocojang commented on src/mocks/data/cart.json:
```diff
@@ -0,0 +1,32 @@
+[
+  {
+    "id": "1",
```
`id`가 두번 관리되는데 어떤 이점이 있을가요?

WaiNaat replied:
위쪽 id는 `cartItemId`이고 아래쪽은 `productId`입니다. msw에서는 구현의 편의를 위해서 두 개를 같게 설정했지만 실제 백엔드 서버에서는 두 id가 다릅니다! 제 생각에는 같은 `productId`를 갖는 서로 다른 `cartItemId`의 개수를 세서 장바구니에 많이 담은 상품을 추적하거나 하는 데 쓸 수 있을 것 같아요. 사용자별 장바구니의 고유성을 유지하는 데 이점이 있을 것 같다는 생각이 듭니다

pocojang commented on src/mocks/data/cart.json:
```diff
@@ -0,0 +1,32 @@
+      "id": 1,
+      "name": "지빠귀",
+      "price": 10000,
+      "imageUrl": "https://images.unsplash.com/photo-1606567595334-d39972c85dbe?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1887&q=80"
```
뒤에 붙은 쿼리도 상수로 관리해야할까요?

WaiNaat replied:
앗 생각없이 사진 우클릭해서 주소복사했더니 이런 문제가 생긴 것 같아요...

pocojang reviewed (changes requested):
우와~~~반가워요 @WaiNaat 👋 🐦 

## 피드백

### 비동기 코드

정말.. 상당히 혼란합니다.
비동기 코드가 여기저기 숨겨져있네요.. 컴포넌트, Hooks, Recoil.. 여기저기 비동기 코드가 중복 코드로 엄청나게 벌려져있습니다ㅠㅠ
이거 어떻게 수습할 수 있을까요? 😱 

### msw

기똥차게 챙기셨네요..!
그리고 재미있다고 하시는거보니까 대단하십니다.
그냥 칭찬드리고 싶었어요

### 컴포넌트 혼란

컴포넌트가 너무 많은 일을 합니다..!
너무나 많은 코드가 있어요.
한번 모든 컴포넌트 코드를 보고 컴포넌트가 꼭 알아야할 코드와 아닌 코드가 무엇일까 고민해보시는게 좋겠어요 💭 

또한 컴포넌트 네이밍을 다루는 규칙과 계층을.. 잘 구성해보셨으면 좋겠습니다!

### optional

`optional`한 정보를 다룰때 `null`을 상당히 많이 사용하시는데..
`null`, `TS Optional`이 엄청 혼재되어있고 타이핑이나 조건절도 실수가 계속 나오네요.
이 부분 주의하셔야겠습니다.

### server

클라이언트와 서버 그리고 개발 환경과 사용자 환경 분기를 신경쓰시면 좋겠습니다!

### ErrorBoundary

와우.. 바운더리가 엄청나게 많아서 깜짝 놀랐어요.
그걸 장인정신을 발휘해서 페이지마다 또 만드신걸보고 감동 받았습니다.
다만..! DRY!! 불필요한 반복을 줄여야하지 않을까요?
계속 횡단 관심사가 생기지 않던가요? 그 부분 빨리 뽑아내는게 어떨까요?

### Extends

`ErrorBoundary` 컴포넌트를 확장하면서 `extends`를 사용하셨는데요
React에서는 지양하는 것을 권합니다.
그 이유가 뭘까요?
문서와 다양한 자료를 찾아보세요!

[레거시 공식 문서의 Composition vs Inheritance](https://legacy.reactjs.org/docs/composition-vs-inheritance.html)

---

음 전반적으로 백엔드 개발자가 프론트엔드 개발한 느낌을 받았습니다.
코드가 굉장히 어지럽고 혼란한 상태이며 컴포넌트는 정말 뒤죽박죽인 상황인데요.
성장을 위해 컴포넌트에도 많은 관심을 가져주시면 좋겠어요!

물론 요즘 BFF 라는 말도 있듯이 백엔드에 엄청난 강점을 지닌 프론트 개발자분들도 많답니다.
그만큼 스토리북, TDD, MSW에는 정말 다른 크루들보다 압도적으로 많은 신경을 쓰고 관리하신 점을 보면 백엔드 지식에 어느정도 강점이 있는게 참새의 매력이 아닐까 싶기도 하네요.

일단 바쁘시죠..? 정리해야할 코드가 산더미이지만 그래도 훌륭한 앱을 만드셨으니 축하드립니다!
하지만 버그는 꼬옥 고쳐주세요!!
그럼 수정 후 다시 뵐게요 👋

WaiNaat committed:
refactor: msw 켜는 로직 분리

WaiNaat committed:
refactor: 서버 변경 시 msw 켜고 끄는 로직 삭제

- 현재 msw는 배포 단계에서 하나의 백엔드 서버처럼 동작함
- 클라이언트에서 서버와의 연결을 끊더라도 서버가 꺼지지는 않으므로
서버 변경 시 msw를 끄는 로직은 불필요하다고 판단

WaiNaat committed:
refactor: type only import 적용

WaiNaat committed:
refactor: 이벤트 핸들러 네이밍 변경

WaiNaat committed:
refactor: 색상값 상수화

WaiNaat committed:
refactor: 에러바운더리 파일명 변경

컴포넌트명에 맞게 변경

WaiNaat committed:
refactor: 불필요한 null 삭제

WaiNaat committed:
refactor: 불필요한 effect 삭제

- 서버 변경 시 메인 페이지로 리다이렉트되므로 불필요한 이펙트

WaiNaat committed:
refactor: 서버 변경 시 주문 상세 정보 상태 초기화

WaiNaat committed:
feat: 주문 상세 보기 없이 페이지 접근 시 에러 처리

WaiNaat committed:
refactor: 단독 navigate는 Link 태그로 변경

WaiNaat committed:
refactor: 불필요한 핸들러 함수 생성 제거

WaiNaat committed:
refactor: mock data imageUrl 불필요한 쿼리 삭제

WaiNaat committed:
fix: 주문 목록 맨 밑에 숫자 나오는 버그 수정

WaiNaat committed:
refactor: 조건문 간결하게 변경

WaiNaat committed:
refactor: 타입 named export로 변경

WaiNaat committed:
chore: CRA와 중복되는 eslint 설정 제거

WaiNaat reviewed (commented):
코드 리뷰 감사합니다 포코!!

코멘트를 읽으면서 제 코드를 다시 보니까 제가 얼마나 구현에 급급했는지 느껴지더라구요... 많이 반성했습니다.
마음 같아서는 리뷰해주신 부분들 모두 반영해서 다시 요청드리고 싶은데, 방학 기간에 코드 리뷰가 어떻게 진행되는지 잘 몰라서 우선 코멘트 주신 부분들이랑 비동기만 수습해서 돌아오겠습니다!

WaiNaat committed:
refactor: 장바구니 fetch 로직 분리

WaiNaat committed:
refactor: 주문 fetch 로직 분리

WaiNaat committed:
refactor: 상품 목록, 주문 fetch api 분리

WaiNaat committed:
refactor: useFetch를 usePromise로 변경

WaiNaat committed:
refactor: usePromise 훅 디렉토리 변경

WaiNaat committed:
refactor: 장바구니 추가 api 함수 불필요한 인자 삭제

WaiNaat committed:
refactor: 컴포넌트에서 fetch api 호출 제거

WaiNaat commented:
포코 안녕하세요 😀

우선 코드에 직접 코멘트로 남겨주신 부분들은 최대한 반영해 보았습니다.
비동기, 컴포넌트, 에러바운더리 상속, 횡단관심사는 다 저한테 정말 도움이 될 것 같아서 꼭 해보고 싶었는데 제대로 된 게 없네요😭

비동기 관련해서는 우선 직접적으로 fetch api를 사용하는 부분은 `src/api/`에 따로 분리를 해 두었습니다. 다른 훅이나 recoil selector에서는 여기에 분리된 함수를 가져와서 사용해요.
recoil 없이 Suspense를 쓰려면 promise를 throw 해야 해서 기존의 useFetch 훅을 usePromise로 더 확장했습니다. 
근데 이 정도로는 극적인 변화가 없는 것 같아서 아쉽습니다...

컴포넌트는 Counter 쪽의 로직을 분리하고 싶었어요. Counter에서 현재 본인 input 값 수정, 전역 상태 수정, api 통신을 다 하고 있어서 이 부분을 세 가지 훅으로 나누면 좋겠다는 생각을 했습니다. 그래서 열심히 시도해 봤는데 잘 안 됐어요. 장바구니에 맨 처음 물건을 넣을 때 response로 오는 `cartItemId`를 무조건 기다려야 하는데 이 부분에서 뭔가 꼬이는 것 같았습니다.

뭔가 시간은 많았던 것 같은데 속도가 안 났던 것 같아 아쉽네요

pocojang commented on src/hooks/api/useCartItemApi.ts:
```diff
@@ -1,13 +1,12 @@
 
-  const addCartItem = async (product: Product) => {
-    const cartItemId = await CartApi.addNewItem(serverName, product);
+  const addCartItem = async (productId: number) => {
```
기존 타입을 재활용하는게 좋겠죠?

pocojang commented on src/hooks/api/usePromise.ts:
```diff
@@ -0,0 +1,39 @@
+import { useCallback, useEffect, useState } from 'react';
+
+const usePromise = <T>(promiseMaker: () => Promise<T>) => {
```
오호... 재미있군요
`TanStack Query` 비스무리한 것을 만들었군요?

pocojang reviewed (approved):
@WaiNaat 👋 

드뎌 레벨2가 끝났군요..!
이전 핑퐁에서 모두 소통했으니 마지막 피드백 간단히 마치도록 하겠습니다.

---

## 피드백

### API 

API 분리 잘 하셨는데 약간 아쉬운게 있어요
`src/api/**.ts`로 분리는 잘 해주셨는데 애초에 이 API 계층에서 데이터까지 다 받은 상태로 다른 곳에서 호출되는 것은 애매한 것 같아요.
그럼 `Getter`로 볼 수 있지 않나 싶기도 하고요.

### 시간 관리

오 신기하게도 시간은 많았는데 속도가 안났다는 재미있는 이야기를 해주셨네요
앞으로는 우선순위를 잘 설정해보세요!

---

레벨3부터는 더 잘하실 수 있을겁니다!!
그럼 이만 💪
